File: tests/integration/service-flow-tests.js
/**
 * Suite de Tests de Flujos - Proyecto Acompañamiento
 * Valida todos los microservicios y flujos de negocio
 * 
 * Rutas de prueba:
 * 1. Health Checks - Validar disponibilidad de todos los servicios
 * 2. Auth Flow - Login, token generation, RBAC
 * 3. Student Management - CRUD operations
 * 4. Teacher Management - CRUD operations
 * 5. Notifications - Send & receive
 * 6. Reports Generation - Create & query reports
 * 7. Analytics - Data aggregation
 * 8. Integration Tests - End-to-end workflows
 */

const axios = require('axios');
const assert = require('assert');

// Configuración de instancias
const INSTANCES = {
  'core': '3.236.99.88',
  'api-gateway': '98.86.94.92',
  'messaging': '35.172.111.207',
  'db': '13.217.220.8',
  'notifications': '98.92.17.165',
  'monitoring': '54.205.158.101',
  'frontend': '52.72.57.10',
  'reportes': '3.236.99.88', // En CORE
  'analytics': '3.236.99.88'  // En CORE
};

// Puertos de servicios
const PORTS = {
  'api-gateway': 8080,
  'auth': 5005,
  'estudiantes': 5002,
  'maestros': 5001,
  'notificaciones': 5006,
  'reportes-estudiantes': 5003,
  'reportes-maestros': 5004,
  'analytics': 5007,
  'monitoring': 5009
};

// Variables globales para test
let authToken = null;
let studentId = null;
let teacherId = null;
let reportId = null;

/**
 * UTILIDADES
 */

class TestRunner {
  constructor() {
    this.results = [];
    this.passed = 0;
    this.failed = 0;
  }

  async run(testName, testFn) {
    try {
      console.log(`\n🧪 Ejecutando: ${testName}`);
      await testFn();
      this.passed++;
      this.results.push({ name: testName, status: '✅ PASS' });
      console.log(`✅ ${testName} - PASSED`);
    } catch (error) {
      this.failed++;
      this.results.push({ name: testName, status: `❌ FAILED: ${error.message}` });
      console.error(`❌ ${testName} - FAILED: ${error.message}`);
    }
  }

  printSummary() {
    console.log('\n' + '='.repeat(80));
    console.log('📊 RESUMEN DE PRUEBAS');
    console.log('='.repeat(80));
    this.results.forEach(r => console.log(`${r.status} - ${r.name}`));
    console.log(`\n✅ Pasadas: ${this.passed}`);
    console.log(`❌ Fallidas: ${this.failed}`);
    console.log(`📊 Total: ${this.passed + this.failed}`);
    console.log(`🎯 Tasa de éxito: ${((this.passed / (this.passed + this.failed)) * 100).toFixed(2)}%`);
    console.log('='.repeat(80) + '\n');
  }
}

const runner = new TestRunner();

/**
 * HEALTH CHECKS - Validar disponibilidad de servicios
 */

async function healthCheck(service, ip, port, path = '/health') {
  const url = `http://${ip}:${port}${path}`;
  try {
    const response = await axios.get(url, { timeout: 5000 });
    assert(response.status === 200, `Expected 200, got ${response.status}`);
    return true;
  } catch (error) {
    throw new Error(`${service} no responde en ${url}: ${error.message}`);
  }
}

async function testHealthChecks() {
  const services = [
    { name: 'API Gateway', ip: INSTANCES['api-gateway'], port: PORTS['api-gateway'] },
    { name: 'Auth Service', ip: INSTANCES['core'], port: PORTS['auth'] },
    { name: 'Estudiantes', ip: INSTANCES['core'], port: PORTS['estudiantes'] },
    { name: 'Maestros', ip: INSTANCES['core'], port: PORTS['maestros'] },
    { name: 'Notificaciones', ip: INSTANCES['notifications'], port: PORTS['notificaciones'] },
    { name: 'Reportes Estudiantes', ip: INSTANCES['core'], port: PORTS['reportes-estudiantes'] },
    { name: 'Reportes Maestros', ip: INSTANCES['core'], port: PORTS['reportes-maestros'] },
    { name: 'Analytics', ip: INSTANCES['core'], port: PORTS['analytics'] }
  ];

  for (const service of services) {
    await runner.run(`Health Check: ${service.name}`, async () => {
      await healthCheck(service.name, service.ip, service.port);
    });
  }
}

/**
 * AUTH FLOW - Login y generación de tokens
 */

async function testAuthLogin() {
  await runner.run('Auth: Login como Maestro', async () => {
    const response = await axios.post(
      `http://${INSTANCES['core']}:${PORTS['auth']}/login`,
      {
        email: 'maestro@test.com',
        password: 'Test@123'
      },
      { timeout: 5000 }
    );

    assert(response.status === 200, `Expected 200, got ${response.status}`);
    assert(response.data.token, 'No token en respuesta');
    authToken = response.data.token;
  });
}

async function testAuthTokenValidation() {
  await runner.run('Auth: Validar token JWT', async () => {
    assert(authToken, 'No hay token para validar');
    
    const response = await axios.post(
      `http://${INSTANCES['core']}:${PORTS['auth']}/validate`,
      { token: authToken },
      { timeout: 5000 }
    );

    assert(response.status === 200, 'Token inválido');
    assert(response.data.valid === true, 'Token no válido');
  });
}

async function testAuthRBAC() {
  await runner.run('Auth: RBAC - Acceso basado en roles', async () => {
    const response = await axios.get(
      `http://${INSTANCES['core']}:${PORTS['auth']}/roles`,
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 200, 'No se obtuvieron roles');
    assert(Array.isArray(response.data.roles), 'Roles no es array');
  });
}

/**
 * STUDENT MANAGEMENT - CRUD Operations
 */

async function testCreateStudent() {
  await runner.run('Estudiantes: CREATE - Crear nuevo estudiante', async () => {
    const response = await axios.post(
      `http://${INSTANCES['api-gateway']}:${PORTS['api-gateway']}/api/estudiantes`,
      {
        nombre: 'Juan Pérez',
        apellido: 'García',
        email: 'juan.perez@test.com',
        grado: '10A',
        estado: 'activo'
      },
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 201 || response.status === 200, `Expected 201/200, got ${response.status}`);
    assert(response.data.id || response.data._id, 'No se retornó ID del estudiante');
    studentId = response.data.id || response.data._id;
  });
}

async function testReadStudent() {
  await runner.run('Estudiantes: READ - Obtener estudiante', async () => {
    assert(studentId, 'No hay studentId para consultar');
    
    const response = await axios.get(
      `http://${INSTANCES['api-gateway']}:${PORTS['api-gateway']}/api/estudiantes/${studentId}`,
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 200, `Expected 200, got ${response.status}`);
    assert(response.data.nombre, 'No tiene nombre el estudiante');
  });
}

async function testUpdateStudent() {
  await runner.run('Estudiantes: UPDATE - Actualizar estudiante', async () => {
    assert(studentId, 'No hay studentId para actualizar');
    
    const response = await axios.put(
      `http://${INSTANCES['api-gateway']}:${PORTS['api-gateway']}/api/estudiantes/${studentId}`,
      { grado: '10B', estado: 'activo' },
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 200, `Expected 200, got ${response.status}`);
  });
}

async function testDeleteStudent() {
  await runner.run('Estudiantes: DELETE - Eliminar estudiante', async () => {
    assert(studentId, 'No hay studentId para eliminar');
    
    const response = await axios.delete(
      `http://${INSTANCES['api-gateway']}:${PORTS['api-gateway']}/api/estudiantes/${studentId}`,
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 200 || response.status === 204, `Expected 200/204, got ${response.status}`);
  });
}

/**
 * TEACHER MANAGEMENT - CRUD Operations
 */

async function testCreateTeacher() {
  await runner.run('Maestros: CREATE - Crear nuevo maestro', async () => {
    const response = await axios.post(
      `http://${INSTANCES['api-gateway']}:${PORTS['api-gateway']}/api/maestros`,
      {
        nombre: 'Carlos López',
        apellido: 'Martínez',
        email: 'carlos.lopez@test.com',
        asignatura: 'Matemáticas',
        estado: 'activo'
      },
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 201 || response.status === 200, `Expected 201/200, got ${response.status}`);
    assert(response.data.id || response.data._id, 'No se retornó ID del maestro');
    teacherId = response.data.id || response.data._id;
  });
}

async function testReadTeacher() {
  await runner.run('Maestros: READ - Obtener maestro', async () => {
    assert(teacherId, 'No hay teacherId para consultar');
    
    const response = await axios.get(
      `http://${INSTANCES['api-gateway']}:${PORTS['api-gateway']}/api/maestros/${teacherId}`,
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 200, `Expected 200, got ${response.status}`);
    assert(response.data.nombre, 'No tiene nombre el maestro');
  });
}

/**
 * NOTIFICATIONS - Send & Receive
 */

async function testSendNotification() {
  await runner.run('Notificaciones: Enviar notificación', async () => {
    const response = await axios.post(
      `http://${INSTANCES['notifications']}:${PORTS['notificaciones']}/api/notificaciones`,
      {
        destinatario: 'test@example.com',
        asunto: 'Test Notification',
        mensaje: 'Este es un mensaje de prueba',
        tipo: 'email'
      },
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 200 || response.status === 201, `Expected 200/201, got ${response.status}`);
    assert(response.data.id || response.data.notificationId, 'No se retornó ID de notificación');
  });
}

async function testGetNotifications() {
  await runner.run('Notificaciones: Obtener notificaciones', async () => {
    const response = await axios.get(
      `http://${INSTANCES['notifications']}:${PORTS['notificaciones']}/api/notificaciones`,
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 200, `Expected 200, got ${response.status}`);
    assert(Array.isArray(response.data) || Array.isArray(response.data.notificaciones), 'No es array de notificaciones');
  });
}

/**
 * REPORTS - Generation & Query
 */

async function testGenerateReport() {
  await runner.run('Reportes: Generar reporte de estudiante', async () => {
    const response = await axios.post(
      `http://${INSTANCES['core']}:${PORTS['reportes-estudiantes']}/api/reportes`,
      {
        studentId: 'test-student-001',
        periodo: '2024-Q1',
        tipo: 'desempeño'
      },
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 200 || response.status === 201, `Expected 200/201, got ${response.status}`);
    assert(response.data.id || response.data.reportId, 'No se retornó ID de reporte');
    reportId = response.data.id || response.data.reportId;
  });
}

async function testQueryReport() {
  await runner.run('Reportes: Consultar reporte generado', async () => {
    assert(reportId, 'No hay reportId para consultar');
    
    const response = await axios.get(
      `http://${INSTANCES['core']}:${PORTS['reportes-estudiantes']}/api/reportes/${reportId}`,
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 200, `Expected 200, got ${response.status}`);
    assert(response.data, 'No se retornó datos del reporte');
  });
}

/**
 * ANALYTICS - Data Aggregation
 */

async function testAnalyticsAggregation() {
  await runner.run('Analytics: Agregar datos de desempeño', async () => {
    const response = await axios.post(
      `http://${INSTANCES['core']}:${PORTS['analytics']}/api/analytics/aggregate`,
      {
        tipo: 'desempeño_estudiantes',
        periodo: '2024-Q1',
        filtros: { grado: '10A' }
      },
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 200, `Expected 200, got ${response.status}`);
    assert(response.data.resultado, 'No hay resultado de agregación');
  });
}

async function testAnalyticsDashboard() {
  await runner.run('Analytics: Obtener dashboard de métricas', async () => {
    const response = await axios.get(
      `http://${INSTANCES['core']}:${PORTS['analytics']}/api/analytics/dashboard`,
      {
        headers: { 'Authorization': `Bearer ${authToken}` },
        timeout: 5000
      }
    );

    assert(response.status === 200, `Expected 200, got ${response.status}`);
    assert(response.data.metricas, 'No hay métricas en dashboard');
  });
}

/**
 * INTEGRATION TESTS - End-to-end workflows
 */

async function testCompleteStudentWorkflow() {
  await runner.run('Integration: Flujo completo de estudiante (CREATE→READ→UPDATE→DELETE)', async () => {
    // CREATE
    const createRes = await axios.post(
      `http://${INSTANCES['api-gateway']}:${PORTS['api-gateway']}/api/estudiantes`,
      {
        nombre: 'Integration Test',
        apellido: 'Student',
        email: 'integration@test.com',
        grado: '11A'
      },
      { headers: { 'Authorization': `Bearer ${authToken}` }, timeout: 5000 }
    );
    assert(createRes.status === 200 || createRes.status === 201);
    const testStudentId = createRes.data.id || createRes.data._id;

    // READ
    const readRes = await axios.get(
      `http://${INSTANCES['api-gateway']}:${PORTS['api-gateway']}/api/estudiantes/${testStudentId}`,
      { headers: { 'Authorization': `Bearer ${authToken}` }, timeout: 5000 }
    );
    assert(readRes.status === 200);

    // UPDATE
    const updateRes = await axios.put(
      `http://${INSTANCES['api-gateway']}:${PORTS['api-gateway']}/api/estudiantes/${testStudentId}`,
      { grado: '11B' },
      { headers: { 'Authorization': `Bearer ${authToken}` }, timeout: 5000 }
    );
    assert(updateRes.status === 200);

    // DELETE
    const deleteRes = await axios.delete(
      `http://${INSTANCES['api-gateway']}:${PORTS['api-gateway']}/api/estudiantes/${testStudentId}`,
      { headers: { 'Authorization': `Bearer ${authToken}` }, timeout: 5000 }
    );
    assert(deleteRes.status === 200 || deleteRes.status === 204);
  });
}

async function testCompleteReportingWorkflow() {
  await runner.run('Integration: Flujo completo de reportes (Generar→Consultar→Exportar)', async () => {
    // GENERATE
    const genRes = await axios.post(
      `http://${INSTANCES['core']}:${PORTS['reportes-estudiantes']}/api/reportes`,
      { studentId: 'integration-001', tipo: 'desempeño' },
      { headers: { 'Authorization': `Bearer ${authToken}` }, timeout: 5000 }
    );
    assert(genRes.status === 200 || genRes.status === 201);
    const testReportId = genRes.data.id || genRes.data.reportId;

    // QUERY
    const queryRes = await axios.get(
      `http://${INSTANCES['core']}:${PORTS['reportes-estudiantes']}/api/reportes/${testReportId}`,
      { headers: { 'Authorization': `Bearer ${authToken}` }, timeout: 5000 }
    );
    assert(queryRes.status === 200);

    // EXPORT
    const exportRes = await axios.get(
      `http://${INSTANCES['core']}:${PORTS['reportes-estudiantes']}/api/reportes/${testReportId}/export?format=pdf`,
      { headers: { 'Authorization': `Bearer ${authToken}` }, timeout: 5000 }
    );
    assert(exportRes.status === 200 || exportRes.status === 201);
  });
}

/**
 * MAIN TEST EXECUTION
 */

async function runAllTests() {
  console.log('\n' + '='.repeat(80));
  console.log('🚀 INICIANDO SUITE DE PRUEBAS DE MICROSERVICIOS');
  console.log('='.repeat(80));

  // 1. Health Checks
  console.log('\n📋 FASE 1: Health Checks');
  await testHealthChecks();

  // 2. Authentication
  console.log('\n📋 FASE 2: Autenticación');
  await testAuthLogin();
  await testAuthTokenValidation();
  await testAuthRBAC();

  // 3. Student Management
  console.log('\n📋 FASE 3: Gestión de Estudiantes');
  await testCreateStudent();
  await testReadStudent();
  await testUpdateStudent();
  await testDeleteStudent();

  // 4. Teacher Management
  console.log('\n📋 FASE 4: Gestión de Maestros');
  await testCreateTeacher();
  await testReadTeacher();

  // 5. Notifications
  console.log('\n📋 FASE 5: Notificaciones');
  await testSendNotification();
  await testGetNotifications();

  // 6. Reports
  console.log('\n📋 FASE 6: Reportes');
  await testGenerateReport();
  await testQueryReport();

  // 7. Analytics
  console.log('\n📋 FASE 7: Analytics');
  await testAnalyticsAggregation();
  await testAnalyticsDashboard();

  // 8. Integration Tests
  console.log('\n📋 FASE 8: Tests de Integración');
  await testCompleteStudentWorkflow();
  await testCompleteReportingWorkflow();

  // Print Summary
  runner.printSummary();
}

// Ejecutar
if (require.main === module) {
  runAllTests().catch(error => {
    console.error('\n❌ Error fatal en suite de pruebas:', error);
    process.exit(1);
  });
}

module.exports = { runner, runAllTests };

File: tests/integration/service_flow_tests.py
#!/usr/bin/env python3
"""
Test Suite de Validación de Microservicios
Proyecto Acompañamiento - Sistema de Gestión Educativa

Ejecuta pruebas completas de todos los servicios y flujos
"""

import requests
import json
import sys
import time
from datetime import datetime
from typing import Dict, List, Tuple, Optional
import subprocess

# Configuración de instancias
INSTANCES = {
    'core': '3.236.99.88',
    'api_gateway': '98.86.94.92',
    'messaging': '35.172.111.207',
    'db': '13.217.220.8',
    'notifications': '98.92.17.165',
    'monitoring': '54.205.158.101',
    'frontend': '52.72.57.10',
}

# Puertos de servicios
PORTS = {
    'api_gateway': 8080,
    'auth': 5005,
    'estudiantes': 5002,
    'maestros': 5001,
    'notificaciones': 5006,
    'reportes_estudiantes': 5003,
    'reportes_maestros': 5004,
    'analytics': 5007,
    'monitoring': 5009,
}

class TestResult:
    def __init__(self, name: str, status: bool, message: str = "", duration: float = 0):
        self.name = name
        self.status = status
        self.message = message
        self.duration = duration
        self.timestamp = datetime.now()

    def __repr__(self):
        icon = "✅" if self.status else "❌"
        return f"{icon} {self.name} ({self.duration:.2f}s) - {self.message}"


class TestSuite:
    def __init__(self):
        self.results: List[TestResult] = []
        self.auth_token: Optional[str] = None
        self.student_id: Optional[str] = None
        self.teacher_id: Optional[str] = None
        self.report_id: Optional[str] = None
        self.timeout = 10

    def run_test(self, name: str, test_fn) -> TestResult:
        """Ejecuta un test y registra el resultado"""
        start = time.time()
        try:
            print(f"🧪 {name}...", end=" ", flush=True)
            test_fn()
            duration = time.time() - start
            result = TestResult(name, True, "OK", duration)
            self.results.append(result)
            print(f"✅ ({duration:.2f}s)")
            return result
        except Exception as e:
            duration = time.time() - start
            result = TestResult(name, False, str(e), duration)
            self.results.append(result)
            print(f"❌ ({duration:.2f}s) - {e}")
            return result

    def print_summary(self):
        """Imprime resumen de resultados"""
        print("\n" + "=" * 100)
        print("📊 RESUMEN DE PRUEBAS")
        print("=" * 100)

        for result in self.results:
            print(result)

        passed = sum(1 for r in self.results if r.status)
        failed = sum(1 for r in self.results if not r.status)
        total = len(self.results)
        success_rate = (passed / total * 100) if total > 0 else 0

        print("\n" + "-" * 100)
        print(f"✅ Pasadas: {passed} | ❌ Fallidas: {failed} | 📊 Total: {total}")
        print(f"🎯 Tasa de éxito: {success_rate:.2f}%")
        print("=" * 100 + "\n")

        return passed, failed

    # ================ HEALTH CHECKS ================

    def health_check(self, service: str, ip: str, port: int, path: str = "/health"):
        """Verifica que un servicio esté disponible"""
        url = f"http://{ip}:{port}{path}"
        response = requests.get(url, timeout=self.timeout)
        assert response.status_code == 200, f"Status {response.status_code}"

    def test_health_checks(self):
        """Ejecuta health checks de todos los servicios"""
        services = [
            ("API Gateway", INSTANCES['api_gateway'], PORTS['api_gateway']),
            ("Auth Service", INSTANCES['core'], PORTS['auth']),
            ("Estudiantes", INSTANCES['core'], PORTS['estudiantes']),
            ("Maestros", INSTANCES['core'], PORTS['maestros']),
            ("Notificaciones", INSTANCES['notifications'], PORTS['notificaciones']),
            ("Reportes Estudiantes", INSTANCES['core'], PORTS['reportes_estudiantes']),
            ("Reportes Maestros", INSTANCES['core'], PORTS['reportes_maestros']),
            ("Analytics", INSTANCES['core'], PORTS['analytics']),
        ]

        for service, ip, port in services:
            self.run_test(
                f"Health: {service}",
                lambda s=service, i=ip, p=port: self.health_check(s, i, p)
            )

    # ================ AUTHENTICATION ================

    def test_auth_login(self):
        """Login de usuario maestro"""
        def auth_login():
            response = requests.post(
                f"http://{INSTANCES['core']}:{PORTS['auth']}/login",
                json={"email": "maestro@test.com", "password": "Test@123"},
                timeout=self.timeout
            )
            assert response.status_code in [200, 201], f"Status {response.status_code}"
            data = response.json()
            assert "token" in data, "No token en respuesta"
            self.auth_token = data["token"]

        self.run_test("Auth: Login maestro", auth_login)

    def test_auth_token_validation(self):
        """Valida token JWT"""
        def token_validation():
            assert self.auth_token, "No hay token para validar"
            response = requests.post(
                f"http://{INSTANCES['core']}:{PORTS['auth']}/validate",
                json={"token": self.auth_token},
                timeout=self.timeout
            )
            assert response.status_code == 200, f"Status {response.status_code}"
            data = response.json()
            assert data.get("valid") == True, "Token no válido"

        self.run_test("Auth: Validar token JWT", token_validation)

    def test_auth_rbac(self):
        """Prueba control de acceso basado en roles"""
        def rbac():
            assert self.auth_token, "No hay token para consultar roles"
            response = requests.get(
                f"http://{INSTANCES['core']}:{PORTS['auth']}/roles",
                headers={"Authorization": f"Bearer {self.auth_token}"},
                timeout=self.timeout
            )
            assert response.status_code == 200, f"Status {response.status_code}"
            data = response.json()
            assert isinstance(data.get("roles", []), list), "Roles no es lista"

        self.run_test("Auth: RBAC - Control de roles", rbac)

    # ================ STUDENT MANAGEMENT ================

    def test_create_student(self):
        """Crea nuevo estudiante"""
        def create():
            response = requests.post(
                f"http://{INSTANCES['api_gateway']}:{PORTS['api_gateway']}/api/estudiantes",
                json={
                    "nombre": "Juan",
                    "apellido": "Test",
                    "email": f"juan.{int(time.time())}@test.com",
                    "grado": "10A",
                },
                headers={"Authorization": f"Bearer {self.auth_token}"},
                timeout=self.timeout
            )
            assert response.status_code in [200, 201], f"Status {response.status_code}"
            data = response.json()
            self.student_id = data.get("id") or data.get("_id")
            assert self.student_id, "No se retornó ID de estudiante"

        self.run_test("Estudiantes: CREATE", create)

    def test_read_student(self):
        """Lee estudiante creado"""
        def read():
            assert self.student_id, "No hay studentId"
            response = requests.get(
                f"http://{INSTANCES['api_gateway']}:{PORTS['api_gateway']}/api/estudiantes/{self.student_id}",
                headers={"Authorization": f"Bearer {self.auth_token}"},
                timeout=self.timeout
            )
            assert response.status_code == 200, f"Status {response.status_code}"

        self.run_test("Estudiantes: READ", read)

    def test_update_student(self):
        """Actualiza estudiante"""
        def update():
            assert self.student_id, "No hay studentId"
            response = requests.put(
                f"http://{INSTANCES['api_gateway']}:{PORTS['api_gateway']}/api/estudiantes/{self.student_id}",
                json={"grado": "10B"},
                headers={"Authorization": f"Bearer {self.auth_token}"},
                timeout=self.timeout
            )
            assert response.status_code == 200, f"Status {response.status_code}"

        self.run_test("Estudiantes: UPDATE", update)

    def test_list_students(self):
        """Lista estudiantes"""
        def list_students():
            response = requests.get(
                f"http://{INSTANCES['api_gateway']}:{PORTS['api_gateway']}/api/estudiantes",
                headers={"Authorization": f"Bearer {self.auth_token}"},
                timeout=self.timeout
            )
            assert response.status_code == 200, f"Status {response.status_code}"
            data = response.json()
            assert isinstance(data, (list, dict)), "Respuesta no es lista o dict"

        self.run_test("Estudiantes: LIST", list_students)

    # ================ TEACHER MANAGEMENT ================

    def test_create_teacher(self):
        """Crea nuevo maestro"""
        def create():
            response = requests.post(
                f"http://{INSTANCES['api_gateway']}:{PORTS['api_gateway']}/api/maestros",
                json={
                    "nombre": "Carlos",
                    "apellido": "Test",
                    "email": f"carlos.{int(time.time())}@test.com",
                    "asignatura": "Matemáticas",
                },
                headers={"Authorization": f"Bearer {self.auth_token}"},
                timeout=self.timeout
            )
            assert response.status_code in [200, 201], f"Status {response.status_code}"
            data = response.json()
            self.teacher_id = data.get("id") or data.get("_id")
            assert self.teacher_id, "No se retornó ID de maestro"

        self.run_test("Maestros: CREATE", create)

    def test_read_teacher(self):
        """Lee maestro creado"""
        def read():
            assert self.teacher_id, "No hay teacherId"
            response = requests.get(
                f"http://{INSTANCES['api_gateway']}:{PORTS['api_gateway']}/api/maestros/{self.teacher_id}",
                headers={"Authorization": f"Bearer {self.auth_token}"},
                timeout=self.timeout
            )
            assert response.status_code == 200, f"Status {response.status_code}"

        self.run_test("Maestros: READ", read)

    def test_list_teachers(self):
        """Lista maestros"""
        def list_teachers():
            response = requests.get(
                f"http://{INSTANCES['api_gateway']}:{PORTS['api_gateway']}/api/maestros",
                headers={"Authorization": f"Bearer {self.auth_token}"},
                timeout=self.timeout
            )
            assert response.status_code == 200, f"Status {response.status_code}"

        self.run_test("Maestros: LIST", list_teachers)

    # ================ NOTIFICATIONS ================

    def test_send_notification(self):
        """Envía notificación"""
        def send():
            response = requests.post(
                f"http://{INSTANCES['notifications']}:{PORTS['notificaciones']}/api/notificaciones",
                json={
                    "destinatario": "test@example.com",
                    "asunto": "Test",
                    "mensaje": "Mensaje de prueba",
                    "tipo": "email"
                },
                headers={"Authorization": f"Bearer {self.auth_token}"},
                timeout=self.timeout
            )
            assert response.status_code in [200, 201], f"Status {response.status_code}"

        self.run_test("Notificaciones: SEND", send)

    def test_get_notifications(self):
        """Obtiene notificaciones"""
        def get_notif():
            response = requests.get(
                f"http://{INSTANCES['notifications']}:{PORTS['notificaciones']}/api/notificaciones",
                headers={"Authorization": f"Bearer {self.auth_token}"},
                timeout=self.timeout
            )
            assert response.status_code == 200, f"Status {response.status_code}"

        self.run_test("Notificaciones: GET", get_notif)

    # ================ REPORTS ================

    def test_generate_report(self):
        """Genera reporte"""
        def generate():
            response = requests.post(
                f"http://{INSTANCES['core']}:{PORTS['reportes_estudiantes']}/api/reportes",
                json={"tipo": "desempeño", "periodo": "2024-Q1"},
                headers={"Authorization": f"Bearer {self.auth_token}"},
                timeout=self.timeout
            )
            assert response.status_code in [200, 201], f"Status {response.status_code}"
            data = response.json()
            self.report_id = data.get("id") or data.get("reportId")

        self.run_test("Reportes: GENERATE", generate)

    def test_query_report(self):
        """Consulta reporte"""
        def query():
            if self.report_id:
                response = requests.get(
                    f"http://{INSTANCES['core']}:{PORTS['reportes_estudiantes']}/api/reportes/{self.report_id}",
                    headers={"Authorization": f"Bearer {self.auth_token}"},
                    timeout=self.timeout
                )
                assert response.status_code == 200, f"Status {response.status_code}"
            else:
                raise Exception("No report ID available")

        if self.report_id:
            self.run_test("Reportes: QUERY", query)

    # ================ ANALYTICS ================

    def test_analytics_metrics(self):
        """Obtiene métricas de analytics"""
        def metrics():
            response = requests.get(
                f"http://{INSTANCES['core']}:{PORTS['analytics']}/api/analytics/metrics",
                headers={"Authorization": f"Bearer {self.auth_token}"},
                timeout=self.timeout
            )
            assert response.status_code == 200, f"Status {response.status_code}"

        self.run_test("Analytics: GET METRICS", metrics)

    # ================ EXECUTE ALL ================

    def run_all(self):
        """Ejecuta toda la suite"""
        print("\n" + "=" * 100)
        print("🚀 INICIANDO SUITE DE PRUEBAS - MICROSERVICIOS")
        print("=" * 100 + "\n")

        print("📋 FASE 1: Health Checks")
        self.test_health_checks()

        print("\n📋 FASE 2: Autenticación")
        self.test_auth_login()
        self.test_auth_token_validation()
        self.test_auth_rbac()

        print("\n📋 FASE 3: Gestión de Estudiantes")
        self.test_create_student()
        self.test_read_student()
        self.test_update_student()
        self.test_list_students()

        print("\n📋 FASE 4: Gestión de Maestros")
        self.test_create_teacher()
        self.test_read_teacher()
        self.test_list_teachers()

        print("\n📋 FASE 5: Notificaciones")
        self.test_send_notification()
        self.test_get_notifications()

        print("\n📋 FASE 6: Reportes")
        self.test_generate_report()
        self.test_query_report()

        print("\n📋 FASE 7: Analytics")
        self.test_analytics_metrics()

        passed, failed = self.print_summary()
        return 0 if failed == 0 else 1


if __name__ == "__main__":
    suite = TestSuite()
    exit_code = suite.run_all()
    sys.exit(exit_code)
